### PR TITLE
Refactor problem; Fix profiles

### LINKED ISSUE #362: Profiles incompatible with fixed parameters (Old: Profiles are per default computed for all including fixed indices)
I think the proper default behavior would be to only compute profiles for free parameters. Opinions?

### PR DESCRIPTION
**Pull request by yannikschaelte**
1) Refactor problem by only storing the _full variables and dynamically create the reduced ones via properties. This will make maintenance easier, the computational overhead should be fine.

2) Fix #362: Correct for fixed parameters when computing profiles.

### CODE REVIEW

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -2
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -2
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -2
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -2
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -2
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -2
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -2
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -2
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
- pypesto/storage/save_to_hdf5.py  1
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -2
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/storage/save_to_hdf5.py  1
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -2
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
- pypesto/storage/save_to_hdf5.py  1
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -3
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
- pypesto/storage/save_to_hdf5.py  1
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -3
         

Clones added
============
- pypesto/problem.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

File: pypesto/problem.py [hunk 1]
```diff
@@ -232,24 +235,18 @@ def unfix_parameters(
...
-        # first clean to be freed indices
-        for i_index, i_parameter in enumerate(parameter_indices):
+        # first clean to-be-freed indices
+        for i_parameter in parameter_indices:
             if i_parameter in self.x_fixed_indices:
+                i_index = self.x_fixed_indices.index(i_parameter)
                 self.x_fixed_indices.pop(i_index)
                 self.x_fixed_vals.pop(i_index)
 
         self.dim = self.dim_full - len(self.x_fixed_indices)
```
Comment (FFroehlich) [inline pypesto/problem.py:245]: duplicate of code in fix_parameters, move to seperate method?

File: pypesto/visualize/profiles.py [hunk 2]
```diff
@@ -120,6 +120,8 @@ def profiles_lowlevel(
         Lower bound.
     ub_full:
         Upper bound.
+    legend_text: List[str]
+        Label for line plots
```
Comment (paulstapor) [inline pypesto/visualize/profiles.py:124]: That docstring is already there in line 115

File: test/test_problem.py [hunk 3]
```diff
@@ -14,7 +15,38 @@ def problem():
...
+    """Test the dynamic problem parameter fixing functions."""
+    problem.fix_parameters(2, 45)
+    assert problem.x_fixed_indices == [0, 1, 5, 2]
+    assert problem.x_fixed_vals == [42, 43, 44, 45]
+    assert problem.x_free_indices == [3, 4, 6, 7, 8, 9]
+
+    problem.unfix_parameters(2)
+    assert problem.x_fixed_indices == [0, 1, 5]
+    assert problem.x_fixed_vals == [42, 43, 44]
+    assert problem.x_free_indices == [2, 3, 4, 6, 7, 8, 9]
```
Comment (paulstapor) [inline test/test_problem.py:45]: :+1:

File: pypesto/problem.py (hunk 1 above)
Comment (paulstapor) [reply pypesto/problem.py:245]: Potentially: move to `self.normalize()` ?

File: pypesto/visualize/profiles.py (hunk 2 above)
Comment (yannikschaelte) [reply pypesto/visualize/profiles.py:124]: yes, that was a merging error. thanks.

**Comment by paulstapor** [comment]
Seems like visualization doesn't work yet with fixed parameters... :/

**Comment by yannikschaelte** [comment]
> Seems like visualization doesn't work yet with fixed parameters... :/

True. This should be adressed here too then.

File: pypesto/problem.py (hunk 1 above)
Comment (yannikschaelte) [reply pypesto/problem.py:245]: both dim and x_free_indices can be properties too. comp. overhead for x_free_indices exists but should be small (O(n_par)).

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/storage/save_to_hdf5.py  1
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/storage/read_from_hdf5.py  -1
+ pypesto/problem.py  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/storage/save_to_hdf5.py  1
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -3
+ pypesto/storage/read_from_hdf5.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
- pypesto/storage/save_to_hdf5.py  1
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -3
+ pypesto/storage/read_from_hdf5.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

**Comment by paulstapor** [comment]
Okay, it seems like the whole story becomes a bit more interesting:
If a parameter is fixed, there will be no profile computed, hence the profilerResult object in the result list is `None`.

It would be easy just to pick those profiles in the profile list, which have been computed and simply plot those. Yet...:
Assume somebody computes a profile for parameter A, and then fixes this parameter. Should the profile still be plotted?
In other words: Do we want to plot all parameter profiles, which have been computed, or restrict furthermore to those, which are not fixed?

**Comment by yannikschaelte** [comment]
> It would be easy just to pick those profiles in the profile list, which have been computed and simply plot those. Yet...:

I would say taking the entries in the result.profile_result[i] that are not None is a good default.

> Assume somebody computes a profile for parameter A, and then fixes this parameter. Should the profile still be plotted?

If the profile exists, it can be plotted as above. And fixing post-hoc should be rather uncommon, I would say ...

> In other words: Do we want to plot all parameter profiles, which have been computed, or restrict furthermore to those, which are not fixed?

I would say the former.

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/storage/save_to_hdf5.py  1
- test/test_problem.py  3
         

Complexity decreasing per file
==============================
+ pypesto/problem.py  -3
+ pypesto/storage/read_from_hdf5.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)

File: pypesto/visualize/profiles.py [hunk 4]
```diff
@@ -53,14 +55,28 @@ def profiles(results, ax=None, profile_indices=None, size=(18.5, 6.5),
...
+        # get parameter indices, for which profiles were computed
+        tmp_indices = [
+            par_id for par_id, prof in
+            enumerate(result.profile_result.list[profile_list_id])
+            if prof is not None]
+        # profile_indices should contain all parameter indices,
+        # for which in at least one of the results a profile exists
+        plottable_indices = list(set().union(plottable_indices,
+                                             tmp_indices))
+
```
Comment (yannikschaelte) [inline pypesto/visualize/profiles.py:70]:
```suggestion
    plottable_indices = set()
    for result in results:
        # get parameter indices, for which profiles were computed
        tmp_indices = [
            par_id for par_id, prof in
            enumerate(result.profile_result.list[profile_list_id])
            if prof is not None]
        # profile_indices should contain all parameter indices,
        # for which in at least one of the results a profile exists
        plottable_indices.update(tmp_indices)
    plottable_indices=sorted(plottable_indices)

```

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_problem.py  3
- pypesto/storage/save_to_hdf5.py  1
         

Complexity decreasing per file
==============================
+ pypesto/storage/read_from_hdf5.py  -1
+ pypesto/problem.py  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5683834&bid=18604937)
